File: golang/tests/animal_test.go
package domain_test

import (
	"testing"

	"claude_tdd/domain"
)

func TestHungryAnimalEatsFood(t *testing.T) {
	tests := []struct {
		species         domain.Species
		food            string
		expectedEmotion domain.Emotion
		expectedEnergy  int
	}{
		{domain.Dog, "bone", domain.EmotionOk, 2},
		{domain.Cat, "fish", domain.EmotionSatisfied, 5},
		{domain.Duck, "shit", domain.EmotionSad, -10},
	}

	for _, tt := range tests {
		t.Run(string(tt.species)+" eating "+tt.food, func(t *testing.T) {
			animal := domain.NewAnimal(tt.species)

			if animal.State() != "hungry" {
				t.Errorf("expected state 'hungry', got '%s'", animal.State())
			}

			animal.Feed(tt.food)

			if animal.Emotion() == nil {
				t.Fatal("expected emotion to be set after feeding")
			}
			if *animal.Emotion() != tt.expectedEmotion {
				t.Errorf("expected emotion '%s', got '%s'", tt.expectedEmotion, *animal.Emotion())
			}
			if animal.Energy() != tt.expectedEnergy {
				t.Errorf("expected energy %d, got %d", tt.expectedEnergy, animal.Energy())
			}
		})
	}
}

func TestDifferentAnimalsMakeDifferentSounds(t *testing.T) {
	tests := []struct {
		species       domain.Species
		expectedSound string
	}{
		{domain.Dog, "Woof"},
		{domain.Cat, "Meow"},
		{domain.Duck, "Quack"},
	}

	for _, tt := range tests {
		t.Run(string(tt.species)+" should make a "+tt.expectedSound+" sound", func(t *testing.T) {
			animal := domain.NewAnimal(tt.species)

			sound := animal.MakeSound()

			if sound != tt.expectedSound {
				t.Errorf("expected sound '%s', got '%s'", tt.expectedSound, sound)
			}
		})
	}
}

File: typescript/tests/domain/Animal.test.ts
import { Animal, AnimalSpecies } from "../../src/domain/Animal";

describe("Animal Behavior", () => {
  describe("A hungry animal eats food", () => {
    it.each([
      { species: "Dog" as AnimalSpecies, food: "bone", emotion: "ok", energy: 2 },
      { species: "Cat" as AnimalSpecies, food: "fish", emotion: "satisfied", energy: 5 },
      { species: "Duck" as AnimalSpecies, food: "shit", emotion: "sad", energy: -10 },
    ])(
      "$species eating $food should feel $emotion and gain $energy energy",
      ({ species, food, emotion, energy }) => {
        const animal = new Animal(species);
        expect(animal.state).toBe("hungry");

        animal.feed();

        expect(animal.emotion).toBe(emotion);
        expect(animal.energy).toBe(energy);
      }
    );
  });

  describe("Different animals make different sounds", () => {
    it.each([
      { species: "Dog" as AnimalSpecies, sound: "Woof" },
      { species: "Cat" as AnimalSpecies, sound: "Meow" },
      { species: "Duck" as AnimalSpecies, sound: "Quack" },
    ])("$species should make a $sound sound", ({ species, sound }) => {
      const animal = new Animal(species);
      expect(animal.makeSound()).toBe(sound);
    });
  });
});
